tests/test_pyglmnet.py

```diff
@@ -164,61 +164,91 @@ def test_glmnet():
-                                   sample=False)
-
-            alpha = 0.
-            reg_lambda = 0.
-            loss_trace = list()
-
-            def callback(beta):
-                Tau = None
-                eta = 2.0
-                group = None
-
-                loss_trace.append(
-                    _loss(distr, alpha, Tau, reg_lambda,
-                          X_train, y_train, eta, group, beta))
-
-            glm = GLM(distr, learning_rate=learning_rate,
-                      reg_lambda=reg_lambda, tol=1e-3, max_iter=5000,
-                      alpha=alpha, solver=solver, score_metric=score_metric,
-                      random_state=random_state, callback=callback)
-            assert(repr(glm))
-
-            glm.fit(X_train, y_train)
-
-            # verify loss decreases
-            assert(np.all(np.diff(loss_trace) <= 1e-7))
-
-            # verify loss at convergence = loss when beta=beta_
-            l_true = _loss(distr, 0., np.eye(beta.shape[0]), 0.,
-                           X_train, y_train, 2.0, None,
-                           np.concatenate(([beta0], beta)))
-            assert_allclose(loss_trace[-1], l_true, rtol=1e-4, atol=1e-5)
-            # beta=beta_ when reg_lambda = 0.
-            assert_allclose(beta, glm.beta_, rtol=0.05, atol=1e-2)
-            betas_.append(glm.beta_)
-
-            y_pred = glm.predict(X_train)
-            assert(y_pred.shape[0] == X_train.shape[0])
-
-        # compare all solvers pairwise to make sure they're close
-        for i, first_beta in enumerate(betas_[:-1]):
-            for second_beta in betas_[i + 1:]:
-                assert_allclose(first_beta, second_beta, rtol=0.05, atol=1e-2)
+        for reg_lambda in reg_lambdas:
+            betas_ = list()
+            for solver in solvers:
+
+                np.random.seed(random_state)
+
+                X_train = np.random.normal(0.0, 1.0, [n_samples, n_features])
+                y_train = simulate_glm(distr, beta0, beta, X_train,
+                                       sample=False)
+                alpha = 0.5
+                loss_trace = list()
+
+                def callback(beta):
+                    Tau = None
+                    eta = 2.0
+                    group = None
+
+                    loss_trace.append(
+                        _loss(distr, alpha, Tau, reg_lambda,
+                              X_train, y_train, eta, group, beta))
+
+                glm = GLM(distr, learning_rate=learning_rate,
+                          reg_lambda=reg_lambda, tol=1e-3, max_iter=5000,
+                          alpha=alpha, solver=solver,
+                          score_metric=score_metric,
+                          random_state=random_state, callback=callback)
+                assert(repr(glm))
+
+                glm.fit(X_train, y_train)
+
+                # verify loss decreases or increases only slightly
+                loss_trace_decreased = np.diff(loss_trace) <= 1e-3
```

Comment:
I suppose an easy check (that would hold for any reg_lambda) would be to count the number of iterations it's non-zero and to assert that it's not more than 2 or so.

Reply:
I'm not sure what makes the most sense, since I'm not extremely familiar with how cdfast should behave. Glancing at loss traces, it seems like increases in loss are not uncommon within cdfast (though when they appear, they're small and usually resolved by the next iteration), in which case a total count would throw false positives when max_iter is high.

Do you think tracking a count of sequential steps where loss increases would make sense? Ex. xgboost's early_stopping_rounds argument where it stops if the loss (on validation) doesn't decrease for that many iterations in a row.

Here since repeated in-bag loss increases signal a bug, I'd think we'd want to monitor those increases and throw an exception within `fit()` itself, which would be natural to integrate into #277 .

Reply:
I think the loss can go up in cdfast but it should not go up in batch (at least for small step sizes). 

early_stopping_rounds sounds like a good idea. However, it might be good to check what's done in the R glmnet package or statsmodel. We may not want to commit too much of the API for one particular solver (cdfast in this case). I know that sklearn uses coordinate descent with cython and in that case you don't have to worry about losses going up.